Ktor JsonFeature ClassNotFoundException

Hi, I added the Kotlin SDK to my app and I'm getting this exception when the app starts:
```
java.lang.NoClassDefFoundError: Failed resolution of: Lio/ktor/client/features/json/JsonFeature$Feature;
                 	at io.ktor.client.features.json.JsonFeature.<clinit>(Unknown Source:0)
                 	at com.telemetrydeck.sdk.TelemetryClient$client$1.invoke(TelemetryClient.kt:19)
                 	at com.telemetrydeck.sdk.TelemetryClient$client$1.invoke(TelemetryClient.kt:18)
                 	at io.ktor.client.HttpClientKt.HttpClient(HttpClient.kt:40)
                 	at com.telemetrydeck.sdk.TelemetryClient.<init>(TelemetryClient.kt:18)
                 	at com.telemetrydeck.sdk.TelemetryManager.send-gIAlu-s(TelemetryManager.kt:74)
                 	at com.telemetrydeck.sdk.TelemetryBroadcastTimer$start$1$1.invokeSuspend(TelemetryBroadcastTimer.kt:48)
                 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                 	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
                 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                 	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@e3b0b42, Dispatchers.IO]
                 Caused by: java.lang.ClassNotFoundException: io.ktor.client.features.json.JsonFeature$Feature
```
I think it is because `JsonFeature` was deprecated in favor of `ContentNegotation` in Ktor now.
note: I'm using Ktor 2.0.0 in my app

--- a/build.gradle
+++ b/build.gradle
@@ -6,7 +6,7 @@ buildscript {
         maven { url 'https://jitpack.io' }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.4.2'
+        classpath 'com.android.tools.build:gradle:8.2.2'
         classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.21'
 
         // NOTE: Do not place your application dependencies here; they belong

--- a/build.gradle
+++ b/build.gradle
@@ -7,7 +7,7 @@ buildscript {
     }
     dependencies {
         classpath 'com.android.tools.build:gradle:8.2.2'
-        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.21'
+        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.8.22'
 
         // NOTE: Do not place your application dependencies here; they belong
         // in the individual module build.gradle files

--- a/gradle.properties
+++ b/gradle.properties
@@ -21,3 +21,6 @@ android.enableJetifier=true
 kotlin.code.style=official
 
 -Dorg.gradle.warning.mode=all
+android.defaults.buildfeatures.buildconfig=true
+android.nonTransitiveRClass=false
+android.nonFinalResIds=false

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.2-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/lib/build.gradle
+++ b/lib/build.gradle
@@ -6,11 +6,11 @@ plugins {
 }
 
 android {
-    compileSdk 31
+    compileSdk 33
 
     defaultConfig {
         minSdk 21
-        targetSdk 31
+        targetSdk 33
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
         consumerProguardFiles "consumer-rules.pro"
@@ -37,7 +37,7 @@ android {
     namespace 'com.telemetrydeck.sdk'
 }
 
-def ktor_version = '1.6.7'
+def ktor_version = '2.3.7'
 def logback_version = '1.2.10'
 def kotlinx_coroutines_version = '1.6.0'
 dependencies {
@@ -50,8 +50,10 @@ dependencies {
     implementation "io.ktor:ktor-client-cio:$ktor_version"
 //    HTTP Serialization
     implementation "io.ktor:ktor-client-serialization:$ktor_version"
-    implementation "org.jetbrains.kotlinx:kotlinx-serialization-properties:1.3.2"
-    implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.3.2"
+    implementation "io.ktor:ktor-client-content-negotiation:$ktor_version"
+    implementation "io.ktor:ktor-serialization-kotlinx-json:$ktor_version"
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.5.1"
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-properties:1.5.1"
 //    HTTP Logging
     implementation "ch.qos.logback:logback-classic:$logback_version"
     implementation "io.ktor:ktor-client-logging:$ktor_version"

--- a/lib/src/main/java/com/telemetrydeck/sdk/EnvironmentMetadataProvider.kt
+++ b/lib/src/main/java/com/telemetrydeck/sdk/EnvironmentMetadataProvider.kt
@@ -67,7 +67,9 @@ class EnvironmentMetadataProvider : TelemetryProvider {
     ): Map<String, String> {
         val signalPayload = additionalPayload.toMutableMap()
         for (item in metadata) {
-            signalPayload[item.key] = item.value
+            if (!signalPayload.containsKey(item.key)) {
+                signalPayload[item.key] = item.value
+            }
         }
         return signalPayload
     }

--- a/lib/src/main/java/com/telemetrydeck/sdk/TelemetryClient.kt
+++ b/lib/src/main/java/com/telemetrydeck/sdk/TelemetryClient.kt
@@ -2,12 +2,15 @@ package com.telemetrydeck.sdk
 
 import io.ktor.client.*
 import io.ktor.client.engine.cio.*
-import io.ktor.client.features.*
-import io.ktor.client.features.json.*
-import io.ktor.client.features.logging.*
+import io.ktor.client.plugins.DefaultRequest
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.client.plugins.logging.DEFAULT
+import io.ktor.client.plugins.logging.LogLevel
+import io.ktor.client.plugins.logging.Logger
+import io.ktor.client.plugins.logging.Logging
 import io.ktor.client.request.*
-import io.ktor.client.statement.*
 import io.ktor.http.*
+import io.ktor.serialization.kotlinx.json.json
 import java.net.URL
 import java.util.*
 
@@ -16,7 +19,9 @@ import java.util.*
  */
 internal class TelemetryClient(private val telemetryAppID: UUID, private val apiBaseURL: URL, private val showDebugLogs: Boolean, private val debugLogger: DebugLogger?) {
     private val client: HttpClient = HttpClient(CIO) {
-        install(JsonFeature)
+        install(ContentNegotiation) {
+            json()
+        }
         if (showDebugLogs && debugLogger != null) {
             install(Logging) {
                 logger = Logger.DEFAULT
@@ -32,10 +37,8 @@ internal class TelemetryClient(private val telemetryAppID: UUID, private val api
     suspend fun send(
         signals: List<Signal>
     ) {
-        val response: HttpResponse = client.request {
-            method = HttpMethod.Post
-            url(getServiceUrl())
-            body = signals
+        val response = client.post(getServiceUrl()) {
+            setBody(signals)
         }
         println(response.status)
         client.close()

--- a/lib/src/test/java/com/telemetrydeck/sdk/EnvironmentMetadataProviderTest.kt
+++ b/lib/src/test/java/com/telemetrydeck/sdk/EnvironmentMetadataProviderTest.kt
@@ -1,0 +1,40 @@
+package com.telemetrydeck.sdk
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import org.junit.Assert
+import org.junit.Rule
+import org.junit.Test
+import java.util.UUID
+
+class EnvironmentMetadataProviderTest {
+    @get:Rule
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+    @Test
+    fun environmentMetadataProvider_sets_client_version() {
+        val appID = "32CB6574-6732-4238-879F-582FEBEB6536"
+        val config = TelemetryManagerConfiguration(appID)
+        val manager =  TelemetryManager.Builder().configuration(config).build(null)
+
+        manager.queue("type", "clientUser", emptyMap())
+
+        val queuedSignal = manager.cache?.empty()?.first()
+
+        Assert.assertNotNull(queuedSignal)
+        Assert.assertEquals(queuedSignal?.payload?.contains("telemetryClientVersion:com.telemetrydeck.sdk"), true)
+    }
+
+    @Test
+    fun environmentMetadataProvider_allows_properties_to_be_set_in_advance() {
+        val appID = "32CB6574-6732-4238-879F-582FEBEB6536"
+        val config = TelemetryManagerConfiguration(appID)
+        val manager =  TelemetryManager.Builder().configuration(config).build(null)
+
+        manager.queue("type", "clientUser", mapOf("telemetryClientVersion" to "my value"))
+
+        val queuedSignal = manager.cache?.empty()?.first()
+
+        Assert.assertNotNull(queuedSignal)
+        Assert.assertEquals(queuedSignal?.payload?.contains("telemetryClientVersion:my value"), true)
+    }
+}

